uast/uast.go

```diff
@@ -262,87 +270,311 @@ type Any interface{}
 // Scope is a temporary definition of a scope semantic type.
 type Scope = Any
 
+// GenNode is a base of every UAST node. It stores positional information.
 type GenNode struct {
 	Positions Positions `json:"@pos,omitempty"`
 }
 
+// Identifier is a name of an entity.
+//
+// What is considered an Identifier:
+// - variable, type, function names;
+// - builtin type names;
+// - package name consisting of a single name element;
+// - goto labels;
+//
+// Not considered an Identifier:
+// - qualified names (see QualifiedIdentifier);
+// - path-like or url-like package names (see String);
 type Identifier struct {
 	GenNode
+	// Name of an entity. Can be any valid UTF8 string.
 	Name string `json:"Name"`
 }
 
+// Roles returns a list of UAST node roles that this node implies.
+func (Identifier) Roles() []role.Role {
+	return []role.Role{
+		role.Identifier,
+	}
+}
+
+// String is an unescaped UTF8 string literal.
+//
+// What is considered a String literal:
+// - escaped string literals;
+// - raw string literals;
+// - path-like or url-like package names;
+//
+// Not considered a String literal:
+// - identifiers (see Identifier);
+// - qualified names (see QualifiedIdentifier);
+// - numeric and boolean literals;
+// - special regexp literals;
 type String struct {
 	GenNode
-	Value  string `json:"Value"`
-	Format string `json:"Format"` // TODO: make an enum later
+	// Value is a UTF8 string literal value.
+	//
+	// Drivers should remove any quotes and unescape the value according to the language rules.
+	Value string `json:"Value"`
+
+	// Format is an optional language-specific string that describes the format of the literal.
+	//
+	// This field can be empty for the most common string literal type of a specific language.
+	// The priority is given to a one-line literal that escapes newline characters.
+	//
+	// TODO: define some well-known formats and maybe make it an enum
+	Format string `json:"Format"`
 }
 
+// QualifiedIdentifier is a name of an entity that consists of multiple simple identifiers,
+// organized in a hierarchy, similar to filesystem paths.
+//
+// What is considered a QualifiedIdentifier:
+// - qualified names that consist of Identifier-like elements;
+//
+// Not considered a QualifiedIdentifier:
+// - path-like or url-like package names (see String);
+// - selector expressions (a->b and a.b in C++);
 type QualifiedIdentifier struct {
 	GenNode
+	// Names is a list of simple identifiers starting from a root level of hierarchy
+	// and ending with leaf identifier.
 	Names []Identifier `json:"Names"`
 }
 
+// Comment is a no-op node that can span multiple lines and provides a human-readable
+// description for code around it.
+//
+// TODO: currently some annotations are also considered a Comment; need to clarify this
 type Comment struct {
 	GenNode
-	Text   string `json:"Text"`
+
+	// Block is set to true for block-style comments.
```

Comment:
What distinguishes a block-style comment in terms of this representation? Both line- and block-comments have text—is the distinction that block:true means there may be line breaks?

Reply:
While writing those comments I also thought that this may be useless for the end user. `Format` should be more useful. Or we can leave it in a token stream if we decide to go that way.